### PR TITLE
[libc++][hardening] Add a bounds check for `valarray` and `bitset`.

### PR DESCRIPTION
**Pull request by var-const**
Add a `valid-element-access` check to `valarray::operator[]` and
`bitset::operator[]`.

### CODE REVIEW

File: libcxx/test/libcxx/numerics/numarray/assert.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+//===----------------------------------------------------------------------===//
```
Comment (var-const) [inline libcxx/test/libcxx/numerics/numarray/assert.pass.cpp:1]: Note: I found it easier to just test all the assertions in a single file (we have precedent in `deque`). We could split by hardening mode later, but I think having a separate test file per hardened function is more trouble than it's worth. No strong feelings, though -- please feel free to push back on this.

File: libcxx/test/libcxx/utilities/template.bitset/assert.abi-v1.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// <bitset>
+
+// Test hardening assertions for std::bitset using ABI v1 (where the const overload of `operator[]` returns
```
Comment (var-const) [inline libcxx/test/libcxx/utilities/template.bitset/assert.abi-v1.pass.cpp:11]: Please let me know if you have any thoughts on how to handle this difference better.

**Review by ldionne — APPROVED** [review]
Some comments. Thanks for adding those!

LGTM with the comments addressed.

**Review by var-const — COMMENTED** [review]
Thank you for the review!

> In practice, the second ABI setting is tested by our CI job that enables the ABI v2.

Do we have a job that tests ABI v2 with hardening enabled? If not, I'm not sure it's worth adding for just this one PR, though, but happy to do so if necessary (ideally in a follow-up :) ).

File: libcxx/docs/Hardening.rst [hunk 3]
```diff
@@ -458,7 +458,7 @@ Hardened containers status
       - Partial
       - N/A
     * - ``bitset``
-      - ❌
+      - ✅
```
Comment (var-const) [inline libcxx/docs/Hardening.rst:461]: I'm still hesitant to list `valarray` as fully hardened given the numerous helper classes -- that would require some investigation!

**Pull request merged**